### ISSUE TITLE
num_procs does not match ps nor pgrep nor grep -l /proc/*

### ISSUE THREAD

**Issue by mozai** [ncabatoff/process-exporter#92]
I'm using the Sept 3rd release, which should be [the latest](https://github.com/ncabatoff/process-exporter/releases/)

I'm getting weird process counts, way more than I believe I should be getting.

config file:  
```
process_names:
- exe: [consul, cron, master, ntp, rsyslogd, sshd]
```
startup:  
`/usr/bin/prometheus-process-exporter -config.path /etc/prometheus/process-exporter.yml -web.listen-address 127.0.0.1:19256`
in use:  
```
curl -s localhost:19256/metrics |grep ^n.*num_procs
namedprocess_namegroup_num_procs{groupname="consul"} 1
namedprocess_namegroup_num_procs{groupname="cron"} 19
namedprocess_namegroup_num_procs{groupname="master"} 3
namedprocess_namegroup_num_procs{groupname="rsyslogd"} 1
namedprocess_namegroup_num_procs{groupname="sshd"} 202

```
this does not match what other tools report:
```
for p in consul cron master rsyslogd sshd; do   echo -n "$p: "; \
  echo -n " pgrep count: " `pgrep -ac $p` ;   \
  echo -n " /proc count " `grep -ailP "\b$p\b" /proc/[1-9]*/cmdline |wc -l`;   \
  echo -n " ps-ef count " `ps -ef |grep -ciP "\b$p\b"`; echo "" ; \
done |column -t
consul:    pgrep  count:  1    /proc  count  1    ps-ef  count  1
cron:      pgrep  count:  5    /proc  count  5    ps-ef  count  5
master:    pgrep  count:  1    /proc  count  1    ps-ef  count  1
rsyslogd:  pgrep  count:  1    /proc  count  1    ps-ef  count  1
sshd:      pgrep  count:  105  /proc  count  105  ps-ef  count  105
```

what I expect:
- the numbers reported by pgrep, ps -ef, and by looking into /proc agree with what is reported by process-exporter

what I see instead:
- three different methods report 5 "cron" processes, process-exporter reports 9 "cron" processes.
- 105 "sshd" processes vs. 202

**Comment by ncabatoff** [comment]
Hi @mozai,

I think you may not be taking into account a key feature of process-exporter: by default, unless -`children=false` is specified, any subprocesses that don't fall into their own groups by virtue of your namegroup configuration are included in their parent process's metrics.  Thus in addition to 105 sshd processes, you have at a minimum the bash (or whatever) subprocesses running underneath them.  Similarly I suspect you have 5 cron daemons and (at scrape time) 4 cron jobs running as children.

**Comment by mozai** [comment]
With `-children=false`, now process_exporter reports 1 sshd num_procs while pgrep, ps -ef|grep and grep /proc/*/cmdline report 98 sshd processes.
I think I understand what it's actually reporting in chidren=true and children=false, it's just surprising.  I was hoping to monitor how many worker processes are launched without also counting the programs launched by the workers and grand-child processes :  sshd -> sshd -> bash -> ps -> grep is counted as either "1 sshd" or "5 sshd" not the 2 actual processes with "sshd" as their argv[0].

Is there a way to count only the processes and all the processes that have a comm that matches the "comm" in the configuration, or only the processes and all the processes with an argv[0] that matches the "exe" in the configuration?

**Comment by ncabatoff** [comment]
> Is there a way to count only the processes and all the processes that have a comm that matches the "comm" in the configuration, 

That should be `comm: [sshd]`.

> or only the processes and all the processes with an argv[0] that matches the "exe" in the configuration?

That should be `exe: [sshd]`.  The thing is, argv[0] gets modified by the sshd children to be something like `sshd: root@pts/0`.  `comm` is what you want for this use case, with `-children=false`.

**Comment by mozai** [comment]
I ended up using `- {comm: [sshd], name: 'sshd'}` to make certain the processes with rewritten {{.ExeBase}}.  Using this with -children=false now I'm getting what other tools are reporting.

Interesting: process_exporter reported "2" cron processes but pgrep and ps|grep only reported one... until I used case-insensitive match.  There was a "cron" process and a "CRON" process simultaneously, so process_exporter was more correct in this case.
